### PR TITLE
dbeaver/dbeaver-ee#1909 Project and resource persmissions support

### PR DESCRIPTION
**Pull request by E1izabeth**
image

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/GlobalPropertyTester.java [hunk 1]
```diff
@@ -61,6 +63,17 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
                 }
                 return false;
             }
+            case PROP_CAN_EDIT_RESOURCE:
+            {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/GlobalPropertyTester.java]:
```suggestion
            case PROP_CAN_EDIT_RESOURCE: {
```

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/GlobalPropertyTester.java [hunk 2]
```diff
@@ -61,6 +63,17 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
...
+                DBPProject project = DBWorkbench.getPlatform().getWorkspace().getActiveProject();
+                return project != null && project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
+            }
+            case PROP_CAN_VIEW_RESOURCE:
+            {
+                DBPProject project = DBWorkbench.getPlatform().getWorkspace().getActiveProject();
+                var x = project != null && project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_VIEW);
+                return x;
+            }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/GlobalPropertyTester.java]:
```suggestion
            case PROP_CAN_VIEW_RESOURCE: {
                DBPProject project = DBWorkbench.getPlatform().getWorkspace().getActiveProject();
                return project != null && project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_VIEW);
            }
```

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java [hunk 3]
```diff
@@ -224,6 +227,9 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
                 return supportsCreatingColumnObject(node, DBSTableIndex.class);
             case PROP_SUPPORTS_CREATING_CONSTRAINT:
                 return supportsCreatingColumnObject(node, DBSEntityConstraint.class);
+            case PROP_PROJECT_RESOURCE_EDITABLE:
+                DBPProject project = node.getOwnerProject();
+                return project != null && project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java]:
```suggestion
            case PROP_PROJECT_RESOURCE_EDITABLE: {
                DBPProject project = node.getOwnerProject();
                return project != null && project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
            }
```

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/datasource/DataSourceToolbarUtils.java [hunk 4]
```diff
@@ -51,20 +56,35 @@ public static DBPDataSourceContainer getCurrentDataSource(IWorkbenchWindow workb
         }
         return null;
     }
+    
+    public static DBPProject getActiveProject(IEditorPart activeEditor, DBPDataSourceContainer dataSource) { 
+        DBPProject activeProject = dataSource == null ? null : dataSource.getProject();
+        if (activeProject == null) {
+            if (activeEditor != null) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/datasource/DataSourceToolbarUtils.java]: Can be merged together

File: plugins/org.jkiss.dbeaver.ui.editors.sql/plugin.xml [hunk 5]
```diff
@@ -311,21 +311,21 @@
             </activeWhen>
             <enabledWhen>
                 <with variable="activePart">
-                    <test property="org.jkiss.dbeaver.core.global.hasActiveProject"/>
+                    <test property="org.jkiss.dbeaver.core.global.canViewResource"/>
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/plugin.xml]:
```suggestion
                    <test property="org.jkiss.dbeaver.core.global.canViewResource"/>
```

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/DataSourcePropertyTester.java [hunk 6]
```diff
@@ -116,8 +125,13 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
...
-                    return contextProvider instanceof DBNDataSource &&
-                        ((DBNDataSource)contextProvider).getOwnerProject().hasRealmPermission(RMConstants.PERMISSION_PROJECT_CONNECTIONS_EDIT);
+                    return resourceProject == null || resourceProject.hasRealmPermission(RMConstants.PERMISSION_PROJECT_CONNECTIONS_EDIT);
+                }
+                case PROP_PROJECT_RESOURCE_EDITABLE: {
+                    return resourceProject == null || resourceProject.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
+                }
+                case PROP_PROJECT_RESOURCE_VIEWABLE: {
+                    return resourceProject == null || resourceProject.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_VIEW);
                 }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/DataSourcePropertyTester.java]: These curly braces not really needed anymore - there are no local variables

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/DataSourcePropertyTester.java [hunk 7]
```diff
@@ -61,9 +65,14 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
             if (!(receiver instanceof DBPContextProvider)) {
                 return false;
             }
-            DBPContextProvider contextProvider = (DBPContextProvider)receiver;
+            DBPContextProvider contextProvider = (DBPContextProvider) receiver;
             @Nullable
             DBCExecutionContext context = contextProvider.getExecutionContext();
+            @Nullable
+            DBPProject resourceProject = receiver instanceof IEditorPart
+                ? EditorUtils.getFileProject(((IEditorPart)receiver).getEditorInput())
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/DataSourcePropertyTester.java]: The `EditorUtils#getFileProject` method can fail and return `null`. It would be better if there was a nullability check.

```suggestion
                ? EditorUtils.getFileProject(((IEditorPart) receiver).getEditorInput())

```

File: plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/EditorUtils.java [hunk 8]
```diff
@@ -73,6 +73,16 @@ public class EditorUtils {
     public static final String COLORS_AND_FONTS_PAGE_ID = "org.eclipse.ui.preferencePages.ColorsAndFonts"; //$NON-NLS-1$
 
     private static final Log log = Log.getLog(EditorUtils.class);
+    
+    public static DBPProject getFileProject(IEditorInput editorInput) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/EditorUtils.java]: Please add nullability annotations

File: plugins/org.jkiss.dbeaver.ui.editors.sql/plugin.xml [hunk 9]
```diff
@@ -749,16 +769,27 @@
...
 
         <menuContribution locationURI="menu:org.jkiss.dbeaver.menu.sql.open">
-            <command commandId="org.jkiss.dbeaver.core.sql.editor.open"/>
-            <command commandId="org.jkiss.dbeaver.core.sql.editor.recent"/>
-            <command commandId="org.jkiss.dbeaver.core.sql.editor.create"/>
+            <command commandId="org.jkiss.dbeaver.core.sql.editor.open">
+                <visibleWhen checkEnabled="true"/>
+            </command>
+            <command commandId="org.jkiss.dbeaver.core.sql.editor.recent">
+                <visibleWhen checkEnabled="true" />
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/plugin.xml]:
```suggestion
                <visibleWhen checkEnabled="true"/>
```

File: plugins/org.jkiss.dbeaver.ui.editors.sql/plugin.xml [hunk 10]
```diff
@@ -749,16 +769,27 @@
...
+            </command>
+            <command commandId="org.jkiss.dbeaver.core.sql.editor.create">
+                <visibleWhen checkEnabled="true" />
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/plugin.xml]:
```suggestion
                <visibleWhen checkEnabled="true"/>
```

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/GlobalPropertyTester.java [hunk 11]
```diff
@@ -52,15 +54,22 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
...
-            case PROP_CAN_CREATE_CONNECTION:
-            {
+            case PROP_CAN_CREATE_CONNECTION: {
                 for (DBPProject project : DBWorkbench.getPlatform().getWorkspace().getProjects()) {
                     if (project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_CONNECTIONS_EDIT)) {
                         return true;
                     }
                 }
                 return false;
             }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/GlobalPropertyTester.java]: These curly braces not really needed anymore - there are no local variables

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java [hunk 12]
```diff
@@ -224,6 +226,9 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
                 return supportsCreatingColumnObject(node, DBSTableIndex.class);
             case PROP_SUPPORTS_CREATING_CONSTRAINT:
                 return supportsCreatingColumnObject(node, DBSEntityConstraint.class);
+            case PROP_PROJECT_RESOURCE_EDITABLE:
+                DBPProject project = node.getOwnerProject();
+                return project == null || project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java]:
```suggestion
            case PROP_PROJECT_RESOURCE_EDITABLE: {
                DBPProject project = node.getOwnerProject();
                return project == null || project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
            }
```

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/datasource/DataSourceToolbarUtils.java [hunk 13]
```diff
@@ -55,16 +59,18 @@ public static DBPDataSourceContainer getCurrentDataSource(IWorkbenchWindow workb
...
             MTrimBar topTrim = ((WorkbenchWindow) window).getTopTrim();
+            boolean showConnectionSelector = false;
+            IEditorPart activeEditor = window.getActivePage().getActiveEditor();
+            DBPDataSourceContainer dataSourceContainer = null;
+            if (activeEditor instanceof IDataSourceContainerProvider) {
+                showConnectionSelector = true;
+                dataSourceContainer = ((IDataSourceContainerProvider) activeEditor).getDataSourceContainer();
+            }
+            DBPProject resourceProj = EditorUtils.getFileProject(activeEditor.getEditorInput());
+            Boolean canChangeConn = resourceProj == null || resourceProj.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/datasource/DataSourceToolbarUtils.java]: Please don't use such abbreviations

```suggestion
            boolean canChangeConn = resourceProj == null || resourceProj.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT);
```

File: plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/datasource/ToolsContextMenuHandler.java [hunk 14]
```diff
@@ -78,7 +79,8 @@ public Object execute(ExecutionEvent event) throws ExecutionException
             }
         }
         menuManager.add(new Separator());
-        {
+        
+        if (part.getSite().getService(ICommandService.class).getCommand(SQLEditorCommands.CMD_SQL_EDITOR_OPEN).isEnabled()) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/ui/actions/datasource/ToolsContextMenuHandler.java]: Use `ActionUtils#isCommandEnabled` instead. It looks prettier and it makes sure that the given command exists.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/navigator/DiagramCreateWizard.java [hunk 15]
```diff
@@ -90,13 +94,16 @@ public void init(IWorkbench workbench, IStructuredSelection selection) {
                 }
             }
         }
-        this.folder = diagramFolder;
+        if (diagramFolder != null) {
+            this.folder = diagramFolder;
+            this.project = diagramFolder == null ? null : DBPPlatformDesktop.getInstance().getWorkspace().getProject(diagramFolder.getProject());
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/navigator/DiagramCreateWizard.java]: `diagramFolder` can't be null here

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/DBIconComposite.java [hunk 16]
```diff
@@ -74,6 +74,10 @@ public DBPImage getBottomRight() {
     public void setBottomRight(DBPImage bottomRight) {
         this.bottomRight = bottomRight;
     }
+    
+    public DBPImage getMain() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/DBIconComposite.java:81]: Please add nullability annotations

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java [hunk 17]
```diff
@@ -144,9 +147,21 @@ public String getNodeDescription() {
...
-    {
+    public DBPImage getNodeIcon() {
+        DBPImage iconImage = this.getResourceNodeIcon();
+
+        DBPProject project = getOwnerProject();
+        if (project != null && !project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT)) {
+            iconImage = new DBIconComposite(iconImage, false, null, null, null, DBIcon.OVER_LOCK);
+        }
+
+        return iconImage;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java:162]:
```suggestion
        DBPProject project = getOwnerProject();
        if (project != null && !project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT)) {
            return new DBIconComposite(iconImage, false, null, null, null, DBIcon.OVER_LOCK);
        }

        return getResourceNodeIcon();
```

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java [hunk 18]
```diff
@@ -224,10 +233,22 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
...
+            case PROP_PROJECT_RESOURCE_VIEWABLE:
+                return nodeProjectHasPermission(node, RMConstants.PERMISSION_PROJECT_RESOURCE_VIEW);
         }
         return false;
     }
 
+    /**
+     * Check whether the owner project of the specified node has required permissions
+     */
+    public static boolean nodeProjectHasPermission(DBNNode node, String permissionName) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java]: Please add nullability annotations

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/DBeaverIcons.java [hunk 19]
```diff
@@ -65,11 +65,18 @@ private static class IconDescriptor {
...
             this.imageDescriptor = imageDescriptor;
         }
+
         IconDescriptor(String id, Image image) {
             this.id = id;
             this.image = image;
             this.imageDescriptor = ImageDescriptor.createFromImage(image);
         }
+
+        IconDescriptor(String id, Image image, ImageDescriptor imageDescriptor) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/DBeaverIcons.java]: Please add nullability annotations

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/OverlayImageDescriptor.java [hunk 20]
```diff
@@ -19,20 +19,26 @@
...
 	private ImageDescriptor[] topLeft, topRight, bottomLeft, bottomRight;
 
     public OverlayImageDescriptor(ImageDescriptor baseImage) {
         this.baseImageData = createCachedImageDataProvider(baseImage);
-        this.imageSize = new Point(baseImageData.getWidth(), baseImageData.getHeight());
+        try {
+            this.imageSize = new Point(baseImageData.getWidth(), baseImageData.getHeight());
+        } catch (Throwable ex) {
+            log.debug(ex);
+        }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/OverlayImageDescriptor.java]: Is this a thing?

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java (hunk 17 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java:162]: I don't like multiple exit points in short methods.

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java:162]: It's better than not using result of method call.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java:162]: I looked closer and I can confirm that I can't do what you suggest. I need the result of `this.getResourceNodeIcon()` method to create composite image (the first argument of DBIconComposite ctor).

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/navigator/DiagramCreateWizard.java (hunk 15 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/navigator/DiagramCreateWizard.java]: It can be null, because `DesktopWorkspaceImpl:getResourceDefaultRoot` may return null. Why don't you think so?

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/OverlayImageDescriptor.java (hunk 20 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/OverlayImageDescriptor.java]: I don't need it anymore, thx

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java (hunk 17 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/navigator/DBNResource.java:162]: Oh, now I see. My apologies for giving inappropriate suggestion.

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java [hunk 21]
```diff
@@ -190,7 +190,15 @@ public void widgetSelected(SelectionEvent e) {
         } else {
             UIUtils.createPlaceholder(childComposite, 1);
         }
-        Button projectCheckbox = UIUtils.createCheckbox(childComposite, SQLEditorMessages.script_selector_project_scripts, false);
+        Button projCheckbox = UIUtils.createCheckbox(childComposite, SQLEditorMessages.script_selector_project_scripts, false);
+        projCheckbox.addSelectionListener(new SelectionAdapter() {
+            @Override
+            public void widgetSelected(SelectionEvent e) {
+            DBWorkbench.getPlatform().getPreferenceStore().setValue(PREF_SCRIPT_SELECTOR_SHOW_PROJECT_SCRIPTS, projCheckbox.getSelection());
+            useProjectScripts(projCheckbox.getSelection());
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]: Please fix indentation

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java [hunk 22]
```diff
@@ -190,7 +190,15 @@ public void widgetSelected(SelectionEvent e) {
         } else {
             UIUtils.createPlaceholder(childComposite, 1);
         }
-        Button projectCheckbox = UIUtils.createCheckbox(childComposite, SQLEditorMessages.script_selector_project_scripts, false);
+        Button projCheckbox = UIUtils.createCheckbox(childComposite, SQLEditorMessages.script_selector_project_scripts, false);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]: Why this change?

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java (hunk 22 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]: Checkstyle was worried about 12 lines between declaration and usage of this checkbox, so I moved it up, but the moved line is too long, so I decided to rename it.

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]: It seems to be only 130 characters while the limit is 140. So you can keep its original name.

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]: I have already fixed it by introducing the local variable for preference store. Line 199 was more than 140 characters.

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]: We are talking about line 193

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]: Please, see latest changes. It's outdated

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java [hunk 23]
```diff
@@ -160,26 +167,40 @@ protected Control createDialogArea(Composite parent) {
...
+        if (this.project.hasRealmPermission(RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT)) {
+            Button newButton = new Button(childComposite, SWT.PUSH | SWT.FLAT);
+            newButton.setText(SQLEditorMessages.script_selector_create_script);
+            newButton.addSelectionListener(new SelectionAdapter() {
+                @Override
+                public void widgetSelected(SelectionEvent e) {
+    
+                    IFile scriptFile;
+                    try {
+                        scriptFile = SQLEditorUtils.createNewScript(
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/controls/ScriptSelectorPanel.java]:
```suggestion
                    try {
                        IFile scriptFile = SQLEditorUtils.createNewScript(
```

File: plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java [hunk 24]
```diff
@@ -129,12 +132,15 @@ public boolean test(Object receiver, String property, Object[] args, Object expe
...
             case PROP_CAN_DELETE: {
                 if (node instanceof DBNDataSource || node instanceof DBNLocalFolder) {
-                    return node.getOwnerProject().hasRealmPermission(RMConstants.PERMISSION_PROJECT_CONNECTIONS_EDIT);
+                    return nodeProjectHasPermission(node, RMConstants.PERMISSION_PROJECT_CONNECTIONS_EDIT);
                 }
 
                 if (DBNUtils.isReadOnly(node)) {
                     return false;
                 }
+                if (node instanceof DBNNodeWithResource && !nodeProjectHasPermission(node, RMConstants.PERMISSION_PROJECT_RESOURCE_EDIT)) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java:141]: Perhaps the `nodeProjectHasPermission` method should care about `node` being an instance of `DBNNodeWithResource`?

Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.navigator/src/org/jkiss/dbeaver/ui/actions/ObjectPropertyTester.java:141]: No, this method is used not only for DBNNodeWithResource.

**Review by ShadelessFox — APPROVED** [review]

**Comment by Matvey16** [comment]
Verified
